src/workspace/mod.rs

```diff
@@ -0,0 +1,1823 @@
+pub use resolver::WorkspaceResolver;
+pub use resolver::WorkspaceResolverCreateError;
+
+#[allow(clippy::disallowed_types)]
+type UrlRc = crate::sync::MaybeArc<Url>;
+
+#[derive(Debug, Clone)]
+pub struct JsrPackageConfig {
+  pub package_name: String,
+  pub member_ctx: WorkspaceMemberContext,
+  pub config_file: ConfigFileRc,
+}
+
+#[derive(Debug, Clone)]
+pub struct NpmPackageConfig {
+  pub package_nv: PackageNv,
+  pub member_ctx: WorkspaceMemberContext,
+  pub package_json: PackageJsonRc,
+}
+
+#[derive(Debug, Clone, Error)]
+pub enum WorkspaceDiagnosticKind {
+  #[error("The '{0}' option can only be specified in the root workspace deno.json file.")]
+  RootOnlyOption(&'static str),
+  #[error("The '{0}' option can only be specified in a workspace member deno.json file and not the root workspace file.")]
+  MemberOnlyOption(&'static str),
+  #[error("The '{name}' member cannot have the same name as the member at '{other_member}'.")]
+  DuplicateMemberName { name: String, other_member: Url },
+  #[error("The 'workspaces' option should be called 'workspace'.")]
+  DeprecatedWorkspacesOption,
+}
+
+impl WorkspaceDiagnosticKind {
+  /// If the diagnostic should cause the process to shut down.
+  pub fn is_fatal(&self) -> bool {
+    match self {
+      Self::RootOnlyOption { .. }
+      | Self::MemberOnlyOption { .. }
+      | Self::DeprecatedWorkspacesOption => false,
+      Self::DuplicateMemberName { .. } => true,
+    }
+  }
+}
+
+#[derive(Debug, Error, Clone)]
+#[error("{}\n    at {}", .kind, .config_url)]
+pub struct WorkspaceDiagnostic {
+  pub kind: WorkspaceDiagnosticKind,
+  pub config_url: Url,
+}
+
+#[derive(Debug, Error)]
+pub enum ResolveWorkspaceMemberError {
+  #[error("Could not find config file for workspace member in '{}'.", .dir_url)]
+  NotFound { dir_url: Url },
+  #[error(transparent)]
+  ConfigReadError(#[from] ConfigFileReadError),
+  #[error(transparent)]
+  PackageJsonReadError(#[from] PackageJsonLoadError),
+  #[error("Workspace member must be nested in a directory under the workspace.\n  Member: {member_url}\n  Workspace: {workspace_url}")]
+  NonDescendant { workspace_url: Url, member_url: Url },
+}
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub struct WorkspaceDiscoverError(Box<WorkspaceDiscoverErrorKind>);
+
+impl WorkspaceDiscoverError {
+  pub fn as_kind(&self) -> &WorkspaceDiscoverErrorKind {
+    &self.0
+  }
+
+  pub fn into_kind(self) -> WorkspaceDiscoverErrorKind {
+    *self.0
+  }
+}
+
+impl<E> From<E> for WorkspaceDiscoverError
+where
+  WorkspaceDiscoverErrorKind: From<E>,
+{
+  fn from(err: E) -> Self {
+    WorkspaceDiscoverError(Box::new(WorkspaceDiscoverErrorKind::from(err)))
+  }
+}
+
+#[derive(Debug, Error)]
+pub enum WorkspaceDiscoverErrorKind {
+  #[error(transparent)]
+  ConfigRead(#[from] ConfigFileReadError),
+  #[error(transparent)]
+  PackageJsonReadError(#[from] PackageJsonLoadError),
+  #[error("Workspace members cannot be empty.\n  Workspace: {0}")]
+  MembersEmpty(Url),
```

Comment:
This seems like a DX hinderance. I think it's fine for the first pass, but what are implication that there are no workspace members?

Reply:
This is https://github.com/denoland/deno_config/pull/54/files

It's just an error if someone specifies `"workspace": []`. They can remove it.